sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala

```diff
@@ -45,6 +46,24 @@ case class CreateHiveTableAsSelectCommand(
   override def innerChildren: Seq[LogicalPlan] = Seq(query)
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
+
+    // relation should move partition columns to the last
+    val (partOutputs, nonPartOutputs) = query.output.partition { a =>
+        tableDesc.partitionColumnNames.contains(a.name)
+    }
+
+    // the CTAS's SELECT partition-outputs order should be consistent with
+    // tableDesc.partitionColumnNames
+    val reorderedPartOutputs = tableDesc.partitionColumnNames.map {
```

Comment:
since we need to reorder the partition columns anyway, how about
```
val partitionAttrs = tableDesc.partitionColumnNames.map { c =>
  query.output.find(_.name == c).getOrElse(throw ...)
}
val partitionSet = AttributeSet(partitionAttrs)
val dataAttrs = query.output.filterNot(partitionSet.contains)
val reorderOutputQuery = Project(dataAttrs ++ partitionAttrs, query)
```